### ISSUE TITLE
ctdTrim doesn't trim equilibration phase very well for some cases

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1104]
I've noticed that for casts where the equilibration phase was at a particular depth, and then the instrument was brought to the surface briefly before beginning the descent (which is good practice, IMHO) that `ctdTrim()` doesn't do a good job of trimming out the equilibration. From the docs I would have expected it to, based on the nlm thing about fitting piecewise linear, but I'm going to dig into it some more. As an example, try running the code below:

```r
library(oce)

p <- c(rep(4, 1000),
       seq(4, 0.5, length.out = 50),
       seq(0.5, 100, length.out=1000),
       rep(100, 100),
       seq(100, 0, length.out=1000))
T <- rnorm(p)
S <- rnorm(p)

d <- as.ctd(S, T, p)

par(mfrow=c(2, 1))
plotScan(d)
plotScan(ctdTrim(d))
```
![image](https://cloud.githubusercontent.com/assets/233584/19630320/c4490170-995e-11e6-99f5-b1fb268f6bf2.png)

**Comment by richardsc** [comment]
Just a note (for myself and potentially for @dankelley), it looks to me like there is a typo at [Line 1903](https://github.com/dankelley/oce/blob/develop/R/ctd.R#L1903) in `ctd.R` and `ctdTrim()`:
```r
            if (length(method) == 1) {
                submethod <- method[1]
                submethod <- "A"
```

Shouldn't it actually be:
```r
            if (length(method) == 1) {
                method <- method[1]
                submethod <- "A"
```
?

**Comment by richardsc** [comment]
I'm going to work on this in the `cr-trim` branch.

**Comment by dankelley** [comment]
Jeeze, when is GH going to make threaded comments.

* Re comment 1: I have been thinking (because of gliders) that we should have a state-space sort of methodology, e.g. I think quite often we could find turn-around points by looking in a $(p, dp/dt)$ space (with some smoothing, most likely)
* Re comment two up: yes, that's a typo, thanks for noticing and maybe you could fix it in your banch.

Again re comment 1:  below is some code of the idea (crude ... it catches the little dip but there could be an additional stage to trim short-lenghted dips or we could smash them with better smoothing)
```R
dt <- 1                                # just for example
p <- c(rep(4, 1000),
       seq(4, 0.5, length.out = 50),
       seq(0.5, 100, length.out=1000),
       rep(100, 100),
       seq(100, 0, length.out=1000))
p <- p + rnorm(length(p), sd=0.01) # 1cm noise
t <- dt * seq_along(p)
dpdt <- diff(smooth(p)) / diff(t)
dpdt <- c(dpdt[1], dpdt) # fake it a bit
fac <- sd(p) / sd(abs(dpdt))
k <- kmeans(cbind(fac*dpdt,p), centers=2) # we'd need to guess centers
par(mfrow=c(1,2))
plot(t, p, pch=20, col=k$cluster)
plot(p, dpdt, pch=20, col=k$cluster)
```
![screen shot 2016-10-23 at 8 41 08 pm](https://cloud.githubusercontent.com/assets/99469/19630455/11bcdb3c-9961-11e6-95c1-e7a488694540.png)

**Comment by richardsc** [comment]
Below is a screenshot from the SBE data processing manual. They have a method for removing surface soak of the kind we're discussing, and I think it would be useful to code something like this method into oce -- perhaps `method="sbe"` or something -- since this is a pretty common surface soak procedure.
![image](https://cloud.githubusercontent.com/assets/233584/19649824/96a38f9a-99dd-11e6-8e4f-c773170381ab.png)

**Comment by dankelley** [comment]
That sounds like a good plan. I think there are likely a lot of `sbe` things though; maybe the arg should be e.g. `method="sbe/soak"` or `"sbe-soak"` or something? (I prefer the `/` notation since that seems a bit more natural; the `-` notation looks like subtraction to me. I guess it could be `sbe:soak`, though). In general, a method could be parameters too, so maybe we need an arg for `method` and another arg for `parameters`?  Figuring out names for things is tricky, sometimes.

**Comment by richardsc** [comment]
Well, this only applies to the method that SBE uses as documented in their manual, so I don't know that it necessarily requires any further distinction. If it did, e.g. if there was more than one SBE trimming method, I think the way to do that would be with the `method=c(method, submethod)` syntax that we have already implemented.

Good point on the parameters though, that is likely the best way to pass things like: `surfaceSoakDepth`, `minimumSoakDepth`, and `maximumSoakDepth` through when calling the "sbe" method for soak trimming.

**Comment by richardsc** [comment]
I've made some good progress for this, in the `cr-trim` branch. There are a couple of examples in  `oce-issues/1104`, including one that uses one of the oce test files from the `CNV` folder (private data). A summary plot of the advantage to the new method is below:
![image](https://cloud.githubusercontent.com/assets/233584/19770438/a0467f62-9c35-11e6-9471-008f5008de4c.png)

There may still be some things to work out (fine-tuning on particular cases), but I'd propose to merge with `develop` soon so that it doesn't go stale. There is no chance of this breaking any existing code since it implements a new `method` to `ctdTrim()`, which has never been used before. If @dankelley agrees I'll plan to merge before the end of the week.

**Comment by dankelley** [comment]
This is looking good. Since it won't break existing code, I agree -- it would be great if you could merge it into `develop`

**Comment by richardsc** [comment]
Merged PR #1106 . Closing.
